node/bindnode/node.go

```diff
@@ -560,16 +560,25 @@ func (w *_assembler) Representation() datamodel.NodeAssembler {
 type basicMapAssembler struct {
 	datamodel.MapAssembler
 
-	builder datamodel.NodeBuilder
-	parent  *_assembler
+	builder   datamodel.NodeBuilder
+	parent    *_assembler
+	converter *converter
 }
 
 func (w *basicMapAssembler) Finish() error {
 	if err := w.MapAssembler.Finish(); err != nil {
 		return err
 	}
 	basicNode := w.builder.Build()
-	w.parent.createNonPtrVal().Set(reflect.ValueOf(basicNode))
+	if w.converter != nil {
+		typ, err := w.converter.customFromAny(basicNode)
+		if err != nil {
+			return err
+		}
+		w.parent.createNonPtrVal().Set(matchSettable(typ, reflect.ValueOf(basicNode)))
+	} else {
+		w.parent.createNonPtrVal().Set(reflect.ValueOf(basicNode))
```

Comment:
is there a way to not need these `reflect.ValueOf` calls on each map operation? is it possible to learn once when the converter is set up / constructed what the type it's going to produce is?

the general goal in how i saw bind-node structure before was that reflection is used when setting up the type system, but once it's set up, the types are fixed, so the actual interaction with data is fast.

Reply:
sadly it's reflection (almost) all the way down; although Dan was able to demonstrate that the perf wasn't too bad because we're limiting mostly to setters and getters with a little bit of inspection (most of that is done at setup, though).

the calls here are taking the value being set, and instantiating it if it's a nil pointer, then setting it to the appropriate form of the value it needs, which requires some finesse - sometimes they're pointers and sometimes not. And of course the `reflect.ValueOf` calls are needed here because `Set()` is a reflection method which can only set reflection value types.